src/hotspot/share/cds/heapShared.hpp

```diff
@@ -282,6 +282,12 @@ class HeapShared: AllStatic {
   static OopHandle _scratch_basic_type_mirrors[T_VOID+1];
   static KlassToOopHandleTable* _scratch_java_mirror_table;
 
+  typedef ResourceHashtable<intptr_t, OopHandle,
+    36137, // prime number
+    AnyObj::C_HEAP,
+    mtClassShared> ResolvedReferenceScratchTable;
```

Comment:
You are using `oop->identity_hash()` as the key for this table. However, it's possible for two `resolved_references` arrays to have the exact same identity. It's better to to use `ResourceHashtable<OopHandle, OopHandle, ...` for this table. Then, you need to define two custom functions for these two parameters for `ResourceHashtable`

```
    unsigned (*HASH)  (K const&),
    bool     (*EQUALS)(K const&, K const&)
```
where the `K` type is `OopHandle`.

The `HASH` function can return `OopHandle::resolve()->identity_hash()` and the `EQUALS` function can compare the values of `OopHandle::resolve()`.

For the coding style, you can search for tables that use `HeapShared::oop_hash` for examples.

Reply:
Actually, using `Oophandle` is too complicated. You can just use the `ConstantPool*` as the key, since each original `resolved_references` array is associated with a unique  `ConstantPool*`.